crates/nargo/src/ops/foreign_calls.rs

```diff
@@ -71,69 +64,88 @@ impl ForeignCall {
+
+    let mut output_strings = Vec::new();
+    let num_values = num_values[0].to_field().to_u128() as usize;
+
+    let mut abi_types = Vec::new();
+    for i in (input_and_abi_values.len() - num_values)..input_and_abi_values.len() {
+        let abi_type = fetch_abi_type(&input_and_abi_values[i])?;
+        abi_types.push(abi_type);
     }
 
-    fn execute_println_format(foreign_call_inputs: &[Vec<Value>]) -> Result<(), ForeignCallError> {
-        // Fetch the message from the first input
-        let (message_as_values, input_and_abi_values) =
-            foreign_call_inputs.split_first().ok_or(ForeignCallError::MissingForeignCallInputs)?;
-        // Fetch the abi type from the foreign call input
-        // The remaining input values should hold what is to be printed
-        let (abi_type, input_values) = fetch_abi_type(input_and_abi_values)?;
-
-        // Fetch the abi field count from the type to account for nested structs
-        let type_size = abi_type.field_count();
-        let input_values_chunks = input_values[0].chunks(type_size as usize);
-
-        let mut output_strings = Vec::new();
-        for input_values in input_values_chunks {
-            let input_values_as_fields = vecmap(input_values, |value| value.to_field());
-            let decoded_value = decode_value(&mut input_values_as_fields.into_iter(), &abi_type)?;
-            let json_value = JsonTypes::try_from_input_value(&decoded_value, &abi_type)?;
-            let output_string = serde_json::to_string(&json_value)
-                .map_err(|err| ForeignCallError::InputParserError(err.into()))?;
-            output_strings.push(output_string);
-        }
+    for i in 0..num_values {
+        let abi_type = &abi_types[i];
+        let type_size = abi_type.field_count() as usize;
 
-        let message_as_fields = vecmap(message_as_values, |value| value.to_field());
-        let message_as_string = decode_string_value(&message_as_fields);
+        let input_values_as_fields = input_and_abi_values[i..(i + type_size)]
+            .iter()
+            .flat_map(|values| vecmap(values, |value| value.to_field()));
 
-        let re = Regex::new(r"\{(\d+)\}").unwrap();
+        let decoded_value = decode_value(&mut input_values_as_fields.into_iter(), abi_type)?;
 
-        let formatted_str = re.replace_all(&message_as_string, |caps: &Captures| {
-            let (_, [target_idx]) = caps.extract();
-            &output_strings[target_idx.parse::<usize>().unwrap()]
-        });
-        println!("{formatted_str}");
-        Ok(())
+        let json_value = JsonTypes::try_from_input_value(&decoded_value, abi_type)?;
+        let output_string = serde_json::to_string(&json_value)
+            .map_err(|err| ForeignCallError::InputParserError(err.into()))?;
+        output_strings.push(output_string);
     }
+
+    let mut output_strings_iter = output_strings.into_iter();
+    let re = Regex::new(r"\{([a-zA-Z0-9]+)\}").unwrap();
```

Comment:
can you put an expect here saying why you expect this to never fail?

Reply:
Done